tests/test_auto_parser.py

```diff
@@ -0,0 +1,61 @@
+import pandas as pd
+import pytest
+from pandas.testing import assert_frame_equal
+
+from fsql.api import read_partitioned_table
+from fsql.column_parser import AutoParser
+from fsql.query import Q_AND, Q_EQ, AtomicQuery
+
+df1 = pd.DataFrame(data={"c1": [0, 1], "c2": ["hello", "world"]})
+df2 = pd.DataFrame(data={"c1": [2, 3], "c2": ["salve", "mundi"]})
+df3 = pd.DataFrame(data={"c1": [4, 5], "c2": ["cthulhu", "rlyeh"]})
+
+
+def test_auto_parser_with_fname(tmp_path):
+    partition1 = tmp_path / "col1=4" / "col2=5" / "colX=a"
+    partition1.mkdir(parents=True)
+    df1.to_json(partition1 / "a1.json", orient="records", lines=True)
+    partition2 = tmp_path / "col1=4" / "col2=6" / "colX=b"
+    partition2.mkdir(parents=True)
+    df2.to_json(partition2 / "a2.json", orient="records", lines=True)
+    partition3 = tmp_path / "col1=9" / "col2=6" / "colX=b"
+    partition3.mkdir(parents=True)
+    df3.to_json(partition3 / "a1.json", orient="records", lines=True)
+
+    # could be replaced with Q_EQ("fname", "a1.json")
+    def query_func(fname: str) -> bool:
+        # matches df1 and df3
+        return fname.startswith("a1")
+
+    query = AtomicQuery(query_func)
+    parser = AutoParser(parse_filenames=True)
+    query = Q_AND(Q_EQ("col1", "4"), query)  # this combination lets only df1 through
+    case2_result = read_partitioned_table(f"file://{tmp_path}/", query, parser)
+
+    case2_expected = df1.assign(col1="4", col2="5", colX="a", fname="a1.json")
+    assert_frame_equal(case2_expected, case2_result)
+
+
+def test_auto_parser_with_fname_broken(tmp_path):
+    partition1 = tmp_path / "col1=4" / "col2" / "colX=a"
+    partition1.mkdir(parents=True)
+    df1.to_json(partition1 / "a1.json", orient="records", lines=True)
+    partition2 = tmp_path / "col1=4" / "col2" / "colX=b"
+    partition2.mkdir(parents=True)
+    df2.to_json(partition2 / "a2.json", orient="records", lines=True)
+    partition3 = tmp_path / "col1=9" / "col2" / "colX=b"
+    partition3.mkdir(parents=True)
+    df3.to_json(partition3 / "a1.json", orient="records", lines=True)
+
+    # could be replaced with Q_EQ("fname", "a1.json")
+    def query_func(fname: str) -> bool:
+        # matches df1 and df3
+        return fname.startswith("a1")
+
+    query = AtomicQuery(query_func)
+    parser = AutoParser(parse_filenames=True)
+    query = Q_AND(Q_EQ("col1", "4"), query)  # this combination lets only df1 through
+
+    # TODO: although the query should match df1 it does not as the partitioning is broken

```

Comment:
nit -- I'd probably replace `broken` with `mixed` in this test case. I believe that even without `parse_filenames`, it fails -- but not due to no object found, but rather, due to unable to split by `=` or smth like that. There are multiple issues when it comes to badly done partitions in general, so I'm not overly ashamed by this particular case. I'd keep this test for sure -- if it suddenly starts to work, or fail for a different reason, it is worth notifying the developer

Reply:
Agree. I've extended tests a bit to see the different ways how to fail 🙂In general, I am fine with it (you are right that mixed structure is not ideal and it's basically the reason for failing in those cases), on the other hand it could be probably confusing for someone.